### ISSUE TITLE
plugins.bloomberg: Always get Could not Find JSON data error from today

### ISSUE THREAD

**Issue by chaoyic** [streamlink/streamlink#4918]
### Checklist

- [X] This is a plugin issue and not a different kind of issue
- [X] [I have read the contribution guidelines](https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#contributing-to-streamlink)
- [X] [I have checked the list of open and recently closed plugin issues](https://github.com/streamlink/streamlink/issues?q=is%3Aissue+label%3A%22plugin+issue%22)
- [X] [I have checked the commit log of the master branch](https://github.com/streamlink/streamlink/commits/master)

### Streamlink version

Latest stable release

### Description

Getting [plugins.bloomberg][error] Could not find JSON data. Invalid URL or bot protection... consistently. 
Used to be able to solve this buy trying VPN or a different network. This does not help now.

My guess is that Bloomberg updated something on their side and the plugin is now broken?
Or Bloomberg just update their "are you a robot" check?

Tried changing User-Agent header manually and it did not help. Can someone point out what is the issue here?

### Debug log

```text
PS C:\Users\***> streamlink -l debug https://www.bloomberg.com/live/us
[cli][debug] OS:         Windows 10
[cli][debug] Python:     3.10.8
[cli][debug] Streamlink: 5.0.1
[cli][debug] Dependencies:
[cli][debug]  isodate: 0.6.1
[cli][debug]  lxml: 4.9.1
[cli][debug]  pycountry: 22.3.5
[cli][debug]  pycryptodome: 3.15.0
[cli][debug]  PySocks: 1.7.1
[cli][debug]  requests: 2.28.1
[cli][debug]  websocket-client: 1.4.1
[cli][debug] Arguments:
[cli][debug]  url=https://www.bloomberg.com/live/us
[cli][debug]  --loglevel=debug
[cli][info] Found matching plugin bloomberg for URL https://www.bloomberg.com/live/us
[plugins.bloomberg][error] Could not find JSON data. Invalid URL or bot protection...
error: No playable streams found on this URL: https://www.bloomberg.com/live/us
```

**Comment by bastimeyer** [comment]
They've updated their bot detection. I had a quick look at this, but I haven't managed to figure this out yet.

Only one HTTP request needs to be made in order to retrieve the JSON data. Chromium directly receives a "200 OK" response with the regular page which contains the JSON data. Copying the request made by Chromium as curl parameters on the other hand results in the "Are you a robot?" page response, for both HTTP/1.1 and HTTP/2, regardless whether consent cookies are sent or not. Since this is the initial HTTP request, any other HTTP request or disabling/enabling JS is irrelevant.

Here's the verbose curl output, and as you can see, the response is "307 s2s_high_score", with a redirection towards `/tosv2.html?vid=&uuid=SOME-UUID&url=SOME-BASE64`.

No idea what Chromium does do differently which they can see and handle.

```
$ curl 'https://www.bloomberg.com/live/us' \
        -H 'authority: www.bloomberg.com' \
        -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9' \
        -H 'accept-language: en-US,en;q=0.9' \
        -H 'cache-control: no-cache' \
        -H 'dnt: 1' \
        -H 'pragma: no-cache' \
        -H 'sec-ch-ua: "Chromium";v="107", "Not=A?Brand";v="24"' \
        -H 'sec-ch-ua-mobile: ?0' \
        -H 'sec-ch-ua-platform: "Linux"' \
        -H 'sec-fetch-dest: document' \
        -H 'sec-fetch-mode: navigate' \
        -H 'sec-fetch-site: none' \
        -H 'sec-fetch-user: ?1' \
        -H 'upgrade-insecure-requests: 1' \
        -H 'user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36' \
        --compressed \
        --http1.1 \
        -v \
        -L
*   Trying 151.101.1.73:443...
* Connected to www.bloomberg.com (151.101.1.73) port 443 (#0)
* ALPN: offers http/1.1
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: none
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN: server accepted http/1.1
* Server certificate:
*  subject: C=US; ST=New York; L=New York; O=Bloomberg LP; CN=www.bloomberg.com
*  start date: May 26 00:00:00 2022 GMT
*  expire date: Mar 24 23:59:59 2023 GMT
*  subjectAltName: host "www.bloomberg.com" matched cert's "www.bloomberg.com"
*  issuer: C=US; O=DigiCert Inc; CN=DigiCert TLS RSA SHA256 2020 CA1
*  SSL certificate verify ok.
> GET /live/us HTTP/1.1
> Host: www.bloomberg.com
> Accept-Encoding: deflate, gzip, br, zstd
> authority: www.bloomberg.com
> accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
> accept-language: en-US,en;q=0.9
> cache-control: no-cache
> dnt: 1
> pragma: no-cache
> sec-ch-ua: "Chromium";v="107", "Not=A?Brand";v="24"
> sec-ch-ua-mobile: ?0
> sec-ch-ua-platform: "Linux"
> sec-fetch-dest: document
> sec-fetch-mode: navigate
> sec-fetch-site: none
> sec-fetch-user: ?1
> upgrade-insecure-requests: 1
> user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36
> 
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* Mark bundle as not supporting multiuse
< HTTP/1.1 307 s2s_high_score
< Connection: close
< Content-Length: 0
< Server: Varnish
< Retry-After: 0
< Location: https://www.bloomberg.com/tosv2.html?vid=&uuid=0ca9ebe4-5ae6-11ed-85ec-4b704b426a6f&url=L2xpdmUvdXM=
< Accept-Ranges: bytes
< Date: Wed, 02 Nov 2022 19:39:22 GMT
< Set-Cookie: _pxhd=ZfetFH2Xd51vkUR8/ze2t7GNli5A1Dnsn8uo-8ovVMKAW1ZZTRpEbxePRuRGMabUTgYuf0L3asX-ir4MoFyQFg==:pvwX9oLur1piCWMzxdIAf0kOMQhxWRXXt0PtaE9ksvNFMgW0xqtdbzyRxJAiEYq5axLUG2Up-zfANw7Tb6oNvGIEKYnvBotrg18lm74oxbo=; Expires=Fri, 01 Jan 2021 00:00:00 GMT; path=/;
< X-Served-By: cache-hhn4053-HHN
< X-Cache: HIT
< X-Cache-Hits: 0
< Strict-Transport-Security: max-age=31557600
< 
* Closing connection 0
* TLSv1.3 (IN), TLS alert, close notify (256):
* TLSv1.3 (OUT), TLS alert, close notify (256):
* Issue another request to this URL: 'https://www.bloomberg.com/tosv2.html?vid=&uuid=0ca9ebe4-5ae6-11ed-85ec-4b704b426a6f&url=L2xpdmUvdXM='
* Hostname www.bloomberg.com was found in DNS cache
*   Trying 151.101.1.73:443...
* Connected to www.bloomberg.com (151.101.1.73) port 443 (#1)
* ALPN: offers http/1.1
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: none
* SSL re-using session ID
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN: server accepted http/1.1
* Server certificate:
*  subject: C=US; ST=New York; L=New York; O=Bloomberg LP; CN=www.bloomberg.com
*  start date: May 26 00:00:00 2022 GMT
*  expire date: Mar 24 23:59:59 2023 GMT
*  subjectAltName: host "www.bloomberg.com" matched cert's "www.bloomberg.com"
*  issuer: C=US; O=DigiCert Inc; CN=DigiCert TLS RSA SHA256 2020 CA1
*  SSL certificate verify ok.
> GET /tosv2.html?vid=&uuid=0ca9ebe4-5ae6-11ed-85ec-4b704b426a6f&url=L2xpdmUvdXM= HTTP/1.1
> Host: www.bloomberg.com
> Accept-Encoding: deflate, gzip, br, zstd
> authority: www.bloomberg.com
> accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
> accept-language: en-US,en;q=0.9
> cache-control: no-cache
> dnt: 1
> pragma: no-cache
> sec-ch-ua: "Chromium";v="107", "Not=A?Brand";v="24"
> sec-ch-ua-mobile: ?0
> sec-ch-ua-platform: "Linux"
> sec-fetch-dest: document
> sec-fetch-mode: navigate
> sec-fetch-site: none
> sec-fetch-user: ?1
> upgrade-insecure-requests: 1
> user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36
> 
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Connection: keep-alive
< Content-Length: 4595
< Cache-Control: public, max-age=5, private, no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0
< Content-Encoding: gzip
< Content-Type: text/html
< Content-Type: text/html; charset=utf-8
< ETag: W/"6362afcb-2abc"
< Last-Modified: Wed, 02 Nov 2022 17:58:35 GMT
< Server: nginx
< Accept-Ranges: bytes
< Date: Wed, 02 Nov 2022 19:39:22 GMT
< Age: 0
< X-Served-By: cache-hhn4033-HHN
< X-Cache: MISS
< X-Cache-Hits: 0
< X-Timer: S1667417962.104323,VS0,VE97
< Strict-Transport-Security: max-age=31557600
< Link: <https://assets.bwbx.io>; rel=preconnect; crossorigin, <https://assets.bwbx.io>; rel=preconnect, <https://sourcepointcmp.bloomberg.com>; rel=preconnect; crossorigin, <https://tpc.googlesyndication.com        >; rel=preconnect; crossorigin, <https://www.google-analytics.com>; rel=preconnect; crossorigin, <https://images.taboola.com>; rel=preconnect, <https://cdn.taboola.com>; rel=preconnect, <https://cdn.taboola.com>; rel=preconnect; crossorigin, <https://sync.taboola.com>; rel=preconnect, <https://sync.taboola.com>; rel=preconnect; crossorigin, <https://www.googletagmanager.com>; rel=preconnect, <https://www.googletagmanager.com>; rel=preconnect; crossorigin
< Vary: Accept-Encoding
< 
<!doctype html>
<html>
<head>
    <title>Bloomberg - Are you a robot?</title>
...
```

**Comment by bastimeyer** [comment]
Ok, according to @mkbloke the `Accept-Encoding` header is the culprit. :+1:

**Comment by chaoyic** [comment]
Very interesting. I'm just wondering how did we/you find out `Accept-Encoding` is the issue here, is there a general process for finding it out or just pure trial and error?

**Comment by bastimeyer** [comment]
He compared the headers sent by his web browser with those sent by Python's requests/urllib3, and since the custom headers defined in the plugin previously didn't have any effect, he probably continued with the default headers of the HTTPSession object and removed those as well while testing the response.

But [as mentioned](https://github.com/streamlink/streamlink/pull/4919#issuecomment-1301193587), it's not clear why removing the header works. If I remove `--compressed` from curl, then the result is still the same for some reason.

**Comment by mkbloke** [comment]
The first thing I did was to modify the plugin to print the response and see what was returned.  Then I started from the other side to bastimeyer - that is, I started with:
```
curl -v https://www.bloomberg.com/live/us
```
which resulted in no content returned.

Then I added the `User-Agent` header that Streamlink sends by default:
```
curl -v -H 'User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0' https://www.bloomberg.com/live/us
```
and that returned the correct page HTML.

The request headers sent by curl with the above are minimal:
```
> GET /live/us HTTP/2
> Host: www.bloomberg.com
> accept: */*
> user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0
```
I removed the extra headers defined in the plugin and also removed the default `Accept-Encoding` header sent by Python requests, as the working curl request was not sending that, and it worked within the plugin then.

Or in short, what bastimeyer has written above!
